restli-server/src/main/java/com/linkedin/restli/internal/server/response/ResponseUtils.java

```diff
@@ -84,6 +95,209 @@ static Object translateCanonicalKeyToAlternativeKeyIfNeeded(Object canonicalKey,
     }
   }
 
+  public static DataMap fillInDefaultOnRecord(RecordDataSchema schema, DataMap dataMap)
+  {
+    DataMap dataWithDefault = new DataMap(DataMapBuilder.getOptimumHashMapCapacityFromSize(dataMap.size()));
+    dataWithDefault.putAll(dataMap);
+    for (RecordDataSchema.Field field : schema.getFields())
+    {
+      DataSchema dataFieldSchema = field.getType();
+      if (dataFieldSchema.getType() == DataSchema.Type.RECORD)
+      {
+        if (dataMap.containsKey(field.getName()) || field.getDefault() != null)
+        {
+          DataMap originalData = dataMap.containsKey(field.getName()) ? (DataMap) dataMap.get(field.getName()) : (DataMap) field.getDefault();
+          dataWithDefault.put(field.getName(), fillInDefaultOnRecord((RecordDataSchema) dataFieldSchema, originalData));
+        }
+      }
+      else if (dataFieldSchema.getType() == DataSchema.Type.ARRAY)
+      {
+        if (dataMap.containsKey(field.getName()) || field.getDefault() != null)
+        {
+          DataList originalData = dataMap.containsKey(field.getName()) ? (DataList) dataMap.get(field.getName()) : (DataList) field.getDefault();
+          dataWithDefault.put(field.getName(), fillInDefaultOnArray((ArrayDataSchema) dataFieldSchema, originalData));
+        }
+      }
+      else if (dataFieldSchema.getType() == DataSchema.Type.TYPEREF)
+      {
+        if (dataMap.containsKey(field.getName()) || field.getDefault() != null)
+        {
+          Object originalData = dataMap.containsKey(field.getName()) ? dataMap.get(field.getName()) : field.getDefault();
+          dataWithDefault.put(field.getName(), fillInDefaultOnTyperef((TyperefDataSchema) dataFieldSchema, originalData));
+        }
+      }
+      else if (dataFieldSchema.getType() == DataSchema.Type.MAP)
+      {
+        if (dataMap.containsKey(field.getName()) || field.getDefault() != null)
+        {
+          DataMap originalData = dataMap.containsKey(field.getName()) ? (DataMap) dataMap.get(field.getName()) : (DataMap) field.getDefault();
+          dataWithDefault.put(field.getName(), fillInDefaultOnMap((MapDataSchema) dataFieldSchema, originalData));
+        }
+      }
+      else if (dataFieldSchema.getType() == DataSchema.Type.UNION)
+      {
+        if (dataMap.containsKey(field.getName()) || field.getDefault() != null)
+        {
+          DataMap originalData = dataMap.containsKey(field.getName()) ? (DataMap) dataMap.get(field.getName()) : (DataMap) field.getDefault();
+          dataWithDefault.put(field.getName(), fillInDefaultOnUnion((UnionDataSchema) dataFieldSchema, originalData));
+        }
+      }
+      else if (!dataWithDefault.containsKey(field.getName()) && field.getDefault() != null)
+      {
+        dataWithDefault.put(field.getName(), field.getDefault());
+      }
+    }
+    return dataWithDefault;
+  }
+
+  public static DataMap fillInDefaultOnMap(MapDataSchema schema, DataMap dataMap)
```

Comment:
fillInDefaultOnXYZ.. can be combined into fillInDefaultValues with switch case. why do we need separate methods for each? and separate switch case constructs?

Reply:
+1, it seems there is a lot of duplication.

You'd just need one fillDefaultOnData which takes dataSchema and an Object, does the branching based on schema type.

In all other methods, you just have to get the nested schema (eg, field, item or value schema) and call this method again.